[JENKINS-60449] Fix and edge case of OptionalExtension blowing up Jenkins

Vlatombe opened the pull request:
<!-- Comment: 
A great PR typically begins with the line below.
Replace XXXXX with the numeric part of the issue's id you created on JIRA.
Please note that if you want your changes backported into LTS, you will need to create a JIRA ticket for it. Read https://jenkins.io/download/lts/#backporting-process for more.
-->
See [JENKINS-60449](https://issues.jenkins-ci.org/browse/JENKINS-60449).

Includes binaries from https://github.com/ikedam/dependencytest/pull/2

If you have an extension like this
```
@OptionalExtension(requirePlugins = {"dependee"})
public class OptionalDependerExtension {
    private static void foo(Dependee d) {

    }
}
```
and `dependee` is not loaded initially, but you load it dynamically later, this blows up, because when Guice does a method scan and can't resolve `Dependee` because it is not yet in classloader at the time of scanning.

This PR can be reviewed commit by commit.
* Introduce a failing test
* Introduce functional changes fixing the test

<!-- Comment: 
If the issue is not fully described in the ticket, add more information here (justification, pull request links, etc.).

 * We do not require JIRA issues for minor improvements.
 * Bugfixes should have a JIRA issue (backporting process).
 * Major new features should have a JIRA issue reference.
-->

### Proposed changelog entries

* Entry 1: Fix an edge case of loading optional dependencies that cause Jenkins to blow up
* ...

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://jenkins.io/changelog/ -->

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@mention

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

Vlatombe committed:
Add new module optional-depender and a new failing test

* Variant plugin is a dependency of optional-depender

Optional-depender contains the following class

@OptionalExtension(requirePlugins = {"dependee"})
public class OptionalDependerExtension {
      private static void foo(Dependee d) {

      }
}

jglick commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -948,52 +948,50 @@ public void dynamicLoad(File arc, boolean removeExisting, @CheckForNull List<Plu
 
     @Restricted(NoExternalUse.class)
     public void start(List<PluginWrapper> plugins) throws Exception {
-      try (ACLContext context = ACL.as(ACL.SYSTEM)) {
```
For other reviewers: **Hide whitespace changes**

Vlatombe replied:
Yes, there were too many whitespaces inconsistencies in this method, I ended up reformatting it.

Vlatombe replied:
https://github.com/jenkinsci/jenkins/pull/4393/files?w=1

Vlatombe committed:
[JENKINS-60449] Move plugin initialization at the end, keep only one extension refresh

Vlatombe commented:
@jglick Updated the PR with a simpler diff (no more refactoring, and a new failing test instead of updating an existing one)

Vlatombe committed:
Improve logic to detect classloading issues

Inspect all declared constructors/methods/fields recursively to exclude
failing classes from Guice rather than have Guice explode.

This also fix the failing test, but the previous commit is still better
because it avoids an extra Jenkins.get().refreshExtensions() call.

Vlatombe committed:
Only test resolveClass method if the classloader is not the bootstrap one

halkeye reviewed (commented):
This is so over my head, so as a core pr reviewer I'm going to abstain from this one but its cool to try to figure out what's going on

+1 for tests

fcojfernandez commented on core/src/main/java/hudson/PluginManager.java:
```diff
@@ -949,23 +949,6 @@ public void dynamicLoad(File arc, boolean removeExisting, @CheckForNull List<Plu
     @Restricted(NoExternalUse.class)
     public void start(List<PluginWrapper> plugins) throws Exception {
       try (ACLContext context = ACL.as(ACL.SYSTEM)) {
-        Jenkins.get().refreshExtensions();
```
isn't the refresh needed anymore? do you know why?

fcojfernandez reviewed (approved):
IIUC, code looks good to me

oleg-nenashev reviewed (approved):
Looks fine though it would be great to run PCT against the patched version

jglick commented:
> Guice does a method scan and can't resolve `Dependee`

By the way: https://github.com/google/guice/pull/1283
